### DISCUSSION TITLE
How to refresh usercontrol, when a property has been changed from codebhind?

### DISCUSSION THREAD

**Discussion by JTrotta** [AvaloniaUI/Avalonia#15993] [Q&A]
I have a usercontrol with a value property. Whenever it changes, I need to refresh/redesign the usercontrol.
The usercontrol is used by a MainWindow and in XAML, the value property is bound to a property in MainWindow viewmodel. Usercontrol  has no viewmodel.

If the value property is changed by mouse interaction, everything works as expected.
Now I need to change the usercontrol value property form a codebehind class (i.e. a service), and the usercontrol must be refreshed. 
I can change the MinWindow viewmodel property, but the usercontrol property is not changed. If I try to change the usercontrol property I got this error: 
"Call from invalid thread"
So the question is: "how to update usercontrol design, changing one of its property?"

Using MVVM pattern with ReactiveUI

**Comment by kekekeks** [comment]
You probably need to mark your property with AffectsRender or AffectsMeasure like this: https://github.com/AvaloniaUI/Avalonia/blob/40e853d6d86dd9860e60d702a220c103eea1da25/src/Avalonia.Visuals/Visual.cs#L105

**Comment by JTrotta** [comment]
May you write an example? 
I Can't figure out aìhow to use AffectsRender .
Thank you

**Comment by Spiralis** [comment]
I too would like to know how to use this.

**Comment by beemobile4net** [comment]
I can't believe this issue has been open now for 4 years.

Anyway, here's how to use 
```csharp
protected static void AffectsRender<T>(params AvaloniaProperty[] properties)  where T : Visual
```
The method is implemented in `Avalonia.Visual` class. Since it's protected, you will only be able to call it from a class, which inherits from `Avalonia.Visual`, typically, when you're writing your own control (usually by inheriting from `Avalonia.Controls.Primitives.TemplatedControl` or from `Avalonia.Controls.Control` - these both inherit from `Avalonia.Visual`). 

You will typically need to call this method, if the control, which you're writing, contains properties, that Avalonia engine should monitor for value changes and if the engine "sees" the value of those properties changed, it would force a new render cycle on your control, in other words, it would call its Render method, defined also in `Avalonia.Visual`:
```csharp
public virtual void Render(DrawingContext context)
```
And so, the use case is, that you're writing your own control, you plan to perform rendering by overriding the Render(DrawingContext) method and your control contains properties, vaules of which influence the rendering. As mentioned above, the properties (which influence the drawing) must be monitored for value changes and therefore, these have to be properties, which implement mechanism for such monitoring. The mechanism is part of styled properties and direct properties. So the properties, values of which influence the drawing, must be styled properties or direct properties (more on that [here](https://docs.avaloniaui.net/docs/guides/custom-controls/defining-properties) and [here](https://docs.avaloniaui.net/docs/guides/custom-controls/how-to-create-advanced-custom-controls#using-a-directproperty-on-another-class)).

The monitoring for changes should be performed on any (and all) instances of your control, therefore, it suffices, that the `AffectsRender` method is static. You probably also want this mechanism to start working before any instance of your control is actually created and so the ideal place for you to call `AffectsRender` method would be static constructor of your control.

So let's say, you're creating a control called `Ctrl`, which will contain one `StyledProperty`, value of which influences the rendering. Your code may look like this:

```csharp
public class Ctrl : TemplatedControl
{
    static Ctrl()
    {
        AffectsRender<Ctrl>(LinethicknessProperty);
    }

    public static readonly StyledProperty<double> LinethicknessProperty = AvaloniaProperty.Register<Ctrl, double>(
     nameof(Linethickness),
     1.0);

    public double Linethickness
    {
        get => this.GetValue(LinethicknessProperty);
        set => this.SetValue(LinethicknessProperty, value);
    }

    public override void Render(DrawingContext context)
    {
        base.Render(context);
        /// Do some drawing with LinethicknessProperty.
    }

}
```
